app/src/main/java/fr/free/nrw/commons/upload/UploadActivity.kt

```diff
@@ -714,7 +723,23 @@ class UploadActivity : BaseActivity(), UploadContract.View, UploadBaseFragment.C
     }
 
     private fun receiveExternalSharedItems() {
-        uploadableFiles = contributionController!!.handleExternalImagesPicked(this, intent).toMutableList()
+        // immediate limit check logic, for avoiding the delay
+        if (intent.action == Intent.ACTION_SEND_MULTIPLE) {
+            val streamList = intent.getParcelableArrayListExtra<Parcelable>(Intent.EXTRA_STREAM)
+            val fileCount = streamList?.size ?: 0
+
+            if (fileCount > MAX_UPLOAD_COUNT) {
+                // limit exceeded,then show dialog immediately and exit.
+                showLimitExceededDialog(fileCount)
+                uploadableFiles.clear() // Clear the list just to be safe
+                return // exits the function before heavy file processing
```

Comment:
Would it be possible to just truncate the list, letting the user upload 20 of the files?
From the user's point of view, selecting the files may have taken time, maybe they selected 21 instead of 20, in which case just removing 1 would be nicer.
What do you think? :-)